### ISSUE TITLE
Which of the following implementations are more object oriented/encouraged?

### ISSUE THREAD

**Issue by freja-p** [nusCS2113-AY1920S1/forum#75]
Class A is a collection of class B objects.  
Class B is a collection of class C objects.  
There is a section of code that needs an attribute of an object from class C (Either get or set the attribute). Which is a more object oriented/encouraged/industry standard way of accessing this attribute?
Implementation 1:  
The external code calls a method in class A ``a.getCattribute()``, which then returns the attribute or sets the attribute in class C.  
The method ``a.getCattribute()`` is simply calling a method on a B object, ``b.getCattribute()``, then the object of C is just a standard getter.

Implementation 2:  
 The external code calls a method in class A ``a.getBobject()``, which returns a B object. The external code then calls ``b.getCobject()`` which returns a C object. Then the external code calls the getter for the C object.

**Comment by okkhoy** [comment]
A good question 👍 
I will wait for others to discuss before I say something.

**Comment by Lucria** [comment]
I think the answer is Implementation 1. Judging from how you mentioned Class A B and C, it seems that class A does not know the existence of class C. As such, in order to access C, we call a method to access class B, then call a method to access class C.

**Comment by chaojliu** [comment]
I also agree with Implementation 1. For 1, external code only has to access A. However for 2, external code should access A, B and C. Imagine if internal relationships between B and C changes, the method signature a.getCattribute() may still work in 1, but not in 2.

**Comment by cyanoei** [comment]
Hi @okkhoy, so what is your opinion on this? thanks!

**Comment by liujiajun** [comment]
I agree with Implementation 1.

As highlighted by @Lucria, from your description, in Implementation 1, A class only knows B class;  In Implementation 2, A class knows both B and C classes. Thus Implementation B could result in more coupling, because a modification to class B or C could potentially result in change to A. 

Hope this helps.

**Comment by cyanoei** [comment]
Thanks everyone, Implementation 1 is correct. [This part of the textbook](https://nuscs2113-ay1920s1.github.io/website/se-book-adapted/chapters/principles.html#law-of-demeter) answers our question.

**Comment by okkhoy** [comment]
Now you know why I didn't answer it earlier 😜  I will be covering LoD tomorrow 😄
Good that you found the answer and the knowledge behind that yourself!
